### PR TITLE
Add a tool for correcting typos

### PR DESCRIPTION
**Pull request by szkarpinski**
## Category:

**Other** (*e.g. Documentation, Tests, Configuration*)

## Description:

TLDR: Finding and fixing over a hundred (#5194) typos took **11 minutes** with this tool

In this PR I share a tool I created and used to correct some typos. It's based on @klecki's CSpell config, but I added some more precise regexes to limit the scope to documentation and comments only. I imagine we could run this tool every few months to quickly correct  accumulated typos in multiple source files.

### How to use this tool
First, let's detect the spelling mistakes with:
```bash
python3 spelling_tools/spellcheck.py
```
This will run CSpell and create `fixme.spelling` file in the following format:
```
(classificed)->() in "dali/operators/segmentation/random_mask_pixel.cc":1360:34:12  -- Pixels are classificed as foreground either
```
Now you need to go through this file and correct the typos. The words are sorted by frequency, so most typos are at the beginning of the file. To fix a typo change:
```
(classificed)->()
```
to:
```
(classificed)->(classified)
```
If it's not a typo, just leave it as it is.

Now, apply the fixes to the files. Run:
```bash
python3 spelling_tools/apply.py
```
It will read `fixme.spelling` and automagically apply your fixes to the files. Also, it will write the words you didn't fix to `cspell_dicts/cspell_dict.txt`, so they won't be reported anymore.

### Why is it so cool?
1. You don't need to browse individual files at any point
2. The unknown words are sorted by frequency and their occurrences are grouped together in the report, so you can easily skip common false-positives.
3. Creating CSpell dictionary requires no additional effort

### Next steps?
- We can think about adding Github action to run the spellcheck on new PRs (but without blocking the merge).
- We can run this monthly (or every 2-3 months) to see if there are some new typos

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by szkarpinski** [comment]
@JanuszL Oh, I forgot about `tools/`. Of course `tools/spelling` is much better than `spelling_tools`

**Review by JanuszL — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Review by klecki — COMMENTED** [review]
Coverage suggestions

File: tools/spelling/spellcheck.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from common import cspell, Problem, FIXME_FILE
+import json
+import re
+
+cspell_out = json.loads(cspell("--show-context", "--reporter", "@cspell/cspell-json-reporter"))
```
Comment (klecki) [inline tools/spelling/spellcheck.py:19]: Can you `if __name__ == "__main__":` this? And maybe provide some `--help` via argparse?

File: tools/spelling/common.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import re
+from dataclasses import dataclass
+
+
+CWD = os.getcwd()
+CSPELL_DOCKER = "ghcr.io/streetsidesoftware/cspell:latest"
+CSPELL_COMMAND = f"docker run -it -v {CWD}:/workdir {CSPELL_DOCKER} \
+lint --no-exit-code --config tools/spelling/cspell.json"
+FIXME_FILE = "fixme.spelling"
+DICT_FILE = "tools/spelling/cspell_dicts/cspell_dict.txt"
```
Comment (klecki) [inline tools/spelling/common.py:26]: I am biased a bit, but I liked the separation of usernames, layouts and other names into their own dictionaries. I guess it would be hard to keep it up with the automation though :(

File: tools/spelling/common.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def cspell(*args):
+    QUIET_FLAGS = ["--no-progress", "--no-color", "--quiet"]
```
Comment (klecki) [inline tools/spelling/common.py:30]: Why are all flags and files defined globaly, and this one is not?

File: tools/spelling/common.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    word: str
+    fix: str
+    ctx: str
+
+    @staticmethod
+    def parse(message):
+        change_pattern = r"\((?P<word>[^)]+)\)->\((?P<fix>[^)]*)\)"
+        loc_pattern = r'"(?P<file>.+)":(?P<offset>\d+):(?P<line>\d+):(?P<col>\d+)'
+        pattern = rf"{change_pattern}\s+in {loc_pattern}\s+-- (?P<ctx>.*)"
+        m = re.match(pattern, message)
```
Comment (klecki) [inline tools/spelling/common.py:51]: I just assume this works :P

Is it possible to have some of the delimiters appear in the problem word, or as those are special characters it won't happen?

File: tools/spelling/apply.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        with open(file, "w") as f:
+            f.write(new_text)
+
+print(f"Applied {fixed} fixes.")
+
+old_dict = {w.lower() for w in read_lines(DICT_FILE)}
+new_dict = old_dict | skipped
+with open(DICT_FILE, "w") as f:
+    f.write("\n".join(w for w in sorted(new_dict)) + "\n")
+print(f"{len(new_dict) - len(old_dict)} words were added to the dictionary.")
```
Comment (klecki) [inline tools/spelling/apply.py:59]: This also looks like it could be encapsulated in `if __name__ == "__main__":` and given a documentation.

**Review by klecki — COMMENTED** [review]
Some more comments

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+docstrings
+downmix
+downmixing
+downscaling
+dragomir
+dtype
+dtypes
+dumitru
+erhan
+externalsource
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell_dicts/cspell_dict.txt:61]: This list is case-insensitive, so `ExternalSource = externalsource`

File: tools/spelling/cspell.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                "/(DocStr|AddOptionalArg|InputDox|Deprecate)\\(\\s*([^,)]+,\\s+)*R?\"[^\"]*\"/g",
+            ],
+        },
+        {
+            "languageId": "python",
+            "allowCompoundWords": false,
+            "includeRegExpList": [
+                "comments",
+                // Match '''long strings''' after a ':'
+                "/:\\s+(\"\"\"|''')[^\\1]*?\\1/g",
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell.json:60]: This PR is focused on spellchecking the docs, that's why I'm interested in docstrings only. Adding all strings would be good, but might result in a flood of false-positives from some test strings etc. I'll try that and see

File: tools/spelling/spellcheck.py (hunk 1 above)
Comment (szkarpinski) [reply tools/spelling/spellcheck.py:19]: Good point

File: tools/spelling/common.py (hunk 2 above)
Comment (szkarpinski) [reply tools/spelling/common.py:26]: That would be nice, but rather impossible for a simple tool :(

File: tools/spelling/common.py (hunk 3 above)
Comment (szkarpinski) [reply tools/spelling/common.py:30]: No idea, will fix

File: tools/spelling/common.py (hunk 4 above)
Comment (szkarpinski) [reply tools/spelling/common.py:51]: Reported words are letter-only, so this won't happen. And filename is in "", so we're safe there

File: tools/spelling/apply.py (hunk 5 above)
Comment (szkarpinski) [reply tools/spelling/apply.py:59]: That's right, will change that

File: tools/spelling/spellcheck.py (hunk 1 above)
Comment (szkarpinski) [reply tools/spelling/spellcheck.py:19]: Done

File: tools/spelling/apply.py (hunk 5 above)
Comment (szkarpinski) [reply tools/spelling/apply.py:59]: Done

File: tools/spelling/cspell.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    "dictionaries": [
+        "project-words",
+    ],
+    "files": [
+        "dali/**/*.cc",
+        "dali/**/*.cuh",
+        "dali/**/*.h",
+        "dali/**/*.cu",
+        "dali/**/*.py",
+        "docs/**/*.rst",
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell.json:20]: Fixed

File: tools/spelling/cspell.json [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        },
+    ],
+    "languageSettings": [
+        {
+            "languageId": "cpp,c",
+            "allowCompoundWords": false,
+            "includeRegExpList": [
+                "comments",
+                // Match strings in schemas
+                "/(DocStr|AddOptionalArg|InputDox|Deprecate)\\(\\s*([^,)]+,\\s+)*R?\"[^\"]*\"/g",
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell.json]: It works with regular strings because `R` is optional (it's `R?` in the regex). I enforce capturing the string contents only by:
```
\"[^\"]*\"
```
which translates into "quote, then anything but not quote, then a quote again"

[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell.json]: And it works with multiline strings because cspell's regexes work in multiline mode by default and a newline matches `[^\"]` just as any other character

File: tools/spelling/cspell.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell.json]: I changed it to spellcheck all strings in both C++ and Python. There is some noise coming from some random strings (like layouts), but after I disabled spellchecking test files it's manageable.

File: tools/spelling/cspell.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (szkarpinski) [reply tools/spelling/cspell.json]: Ok, now I just spellcheck all strings. One magic regex less :)

File: tools/spelling/common.py (hunk 3 above)
Comment (szkarpinski) [reply tools/spelling/common.py:30]: Ok, I read it again and it makes some sense. Those flags are just an implementation detail of `cspell(...)`, so it makes sense to hide them there

**Review by klecki — APPROVED** [review]

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+sinewaves
+sinograms
+sintel
+slaney
+slaney's
+smem
+sndfile
+softmax
+spanev
+spectogram
```
Comment (klecki) [inline tools/spelling/cspell_dicts/cspell_dict.txt:323]:
```suggestion
spectrogram
```

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+autoserialization
+autoserialize
+autotune
+avformat
+avio
+awolant
+baaba
+bayered
+bboxes
+becasue
```
Comment (klecki) [inline tools/spelling/cspell_dicts/cspell_dict.txt:24]:
```suggestion
because
```

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+begining
```
Comment (klecki) [inline tools/spelling/cspell_dicts/cspell_dict.txt:25]:
```suggestion
beginning
```

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+fhwc
+fileref
+filterbank
+flac
+floordiv
+fndefs
+fpow
+frecord
+frecords
+functio
```
Comment (klecki) [inline tools/spelling/cspell_dicts/cspell_dict.txt:104]:
```suggestion
function
```
?

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+librispeech
+librosa
+librosa's
+libsnd
+libsndfile
+libtar
+liftering
+llvmlite
+lmdb
+loging
```
Comment (klecki) [inline tools/spelling/cspell_dicts/cspell_dict.txt:174]:
```suggestion
logging
```

File: tools/spelling/cspell_dicts/cspell_dict.txt [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+nvutils
+nvvl
+o'shaughnessy's
+odirect
+ofapi
+opencv
+opencv's
+opticalflow
+orelse
+orizontal
```
Comment (klecki) [inline tools/spelling/cspell_dicts/cspell_dict.txt:250]:
```suggestion
horizontal
```

**Review by klecki — COMMENTED** [review]
Some typos? in dictionary.

**Comment by szkarpinski** [comment]
!build
